src/blog/non-backtracking-regexp.md

```diff
@@ -0,0 +1,184 @@
+
+Let's try to understand what's going on when matching `/(a*)*b/` against a smaller subject string, say "aaa". This pattern contains nested quantifiers, so we're asking Irregexp to match a *sequence of sequences* of "a", and then match "b". Clearly there is no match because the subject string does not contain "b". However, `/(a*)*/` matches, and it does so in exponentially many different ways:
+
+```javascript
+"aaa"           "aa", "a"           "aa", ""
+"a", "aa"       "a", "a", "a"       "a", "a", ""
+...
+```
+
+A priori, Irregexp cannot rule out that the failure to match the final `/b/` is due to choosing the wrong way of matching `/(a*)*/`, so it has to try all variants. This problem is known as “exponential” or “catastrophic” backtracking.
+
+## RegExps as automata and bytecode
+
+To understand an alternative algorithm that is immune to catastrophic backtracking, we have to take a quick detour via [automata](https://en.wikipedia.org/wiki/Nondeterministic_finite_automaton). Every regular expression is equivalent to an automaton. For example, the RegExp `/(a*)*b/` above corresponds to the following automaton:
+
+![Automaton corresponding to `/(a*)*b/`](/_img/non-backtracking-regexp/example-automaton.svg)
+  
+Note that the automaton is not uniquely determined by the pattern; the one you see above is the automaton you will get by a mechanical translation process, and it’s the one that’s used inside the V8’s new RegExp engine for `/(a*)*/`.
+The unlabeled edges are epsilon transitions: They don’t consume input. Epsilon transitions are necessary to keep the size of the automaton at around the size of the pattern. Naively eliminating epsilon transitions can result in quadratic increase of the number of transitions.
+Epsilon transitions also allow constructing the automaton corresponding to a RegExp from the following four basic kinds of states:
+  
+![RegExp bytecode instructions](/_img/non-backtracking-regexp/state-types.svg)
+
+Here we only classify the transitions *out* of the state, while the transitions into the state are still allowed to be arbitrary. Automata built from only these kinds of states can be represented as *bytecode programs*, with every state corresponding to an instruction. For example, a state with two epsilon transitions is represented as a `FORK` instruction.
+
+## The backtracking algorithm
+
+Let’s revisit the backtracking algorithm that Irregexp is based upon and describe it in terms of automata. Suppose we’re given a bytecode array `code` corresponding to the pattern and want to `test` whether an `input` matches the pattern. Let’s assume that `code` looks something like this:
+
+```javascript
+const code = [
+  {opcode: "FORK", forkPc: 4},
+  {opcode: "CONSUME", char: “1”},
+  {opcode: "CONSUME", char: “2”},
+  {opcode: "JMP", jmpPc: 6},
+  {opcode: "CONSUME", char: “a”},
+  {opcode: "CONSUME", char: “b”},
+  {opcode: "ACCEPT"}
+];

```

Comment:
Nit: to match Google's JS style guide, like other code examples on v8.dev, do this:

```suggestion
const code = [
  {opcode: 'FORK', forkPc: 4},
  {opcode: 'CONSUME', char: '1'},
  {opcode: 'CONSUME', char: '2'},
  {opcode: 'JMP', jmpPc: 6},
  {opcode: 'CONSUME', char: 'a'},
  {opcode: 'CONSUME', char: 'b'},
  {opcode: 'ACCEPT'}
];
```

Same for the other code snippets

Reply:
Oh sorry, I mixed it up and thought " was recommended instead of '. Should be changed now.